### ISSUE TITLE
Paths must be non-empty strings and can't contain '.', '#', '$', '[', or ']'

### ISSUE THREAD

**Issue by binar1** [invertase/react-native-firebase#4175]
I had some problem in using react-native-firebase when I write any Query this error will be shown ,
So after Some looking of last commits i found out the issue is a validation in index.js 

`if (/[.#$\[\]'?]/g.test(path)) {
      throw new Error(
        "Paths must be non-empty strings and can't contain '.', '#', '$', '[', or ']'",
      );
    }` 

Because First  this Method will be Run
 ```
_syncServerTimeOffset() {
    this.ref('.info/serverTimeOffset').on('value', snapshot => {
      console.log(snapshot.val());
      
      this._serverTimeOffset = snapshot.val();
    });
  } 
```

and it Has => . from its Path so Error will be Shown .

after removing => .  from regular expression  the error will be fixed like this 

```
if (/[#$\[\]'?]/g.test(path)) {
      throw new Error(
        "Paths must be non-empty strings and can't contain '.', '#', '$', '[', or ']'",
      );
    }
```

**Comment by mikehardy** [comment]
@binar1 can you link to the related commit + issue to helps us navigate to the root of the problem? I think I remember the one you are talking about but certainty is valuable

Additionally if you proposed a PR for it that would resolve it almost immediately (and if it's as easy as removing a character you can likely just do it using the GitHub web UI - makes the PR flow very easy)

**Comment by owaisulwara** [comment]
Same issue while calling - database().ref(`online/${userId}`) - in react native firebase.

**Comment by owaisulwara** [comment]
![image](https://user-images.githubusercontent.com/28351659/91664626-fbcd7300-eb09-11ea-8b66-ea8c68cd06f4.png)

**Comment by mikehardy** [comment]
@owaisulwara you may edit directly in node_modules to fix the regex in your project https://github.com/ds300/patch-package will handle patch management / eventual upgrade with a nice warning

If you could post a quick PR we could have this sorted in just an hour or so though :pray:

**Comment by fungilation** [comment]
Problem is with PR #4100 

https://github.com/invertase/react-native-firebase/blob/4df5f87db51b169acbe17b4d7eb7e8fed59b8a66/packages/database/lib/index.js#L55 is run on init in constructor, and that immediately fail validation in the new regex at https://github.com/invertase/react-native-firebase/blob/4df5f87db51b169acbe17b4d7eb7e8fed59b8a66/packages/database/lib/index.js#L79, with the `.` in `.info/serverTimeOffset`

**Comment by mikehardy** [comment]
@fungilation if you propose a PR we can likely fix this today

**Comment by fungilation** [comment]
What should be the fix as what was the validation supposed to be? Doesn't seem like "." should be included in the validation at all, but keep the rest of the characters ('#', '$', '[', or ']') as disallowed?

**Comment by mikehardy** [comment]
@fungilation that seems like the correct analysis

**Comment by mikehardy** [comment]
Okay, we have a PR (@fungilation -> :trophy: !) and it's marked pending merge, just waiting on the E2E tests in CI. if they pass I will merge and release

**Comment by mikehardy** [comment]
This is live now - please check!

 - @react-native-firebase/database@7.5.4

**Comment by fungilation** [comment]
Confirmed, working in [my app](https://medium.com/wonderswipe/rethink-mobile-search-10-100x-faster-introducing-wonderswipe-6f2ff0d0e667)

**Comment by deflexable** [comment]
> This is live now - please check!
> 
> * @react-native-firebase/database@7.5.4


@mikehardy , @fungilation  i'm currently facing the same error on the version below

> "@react-native-firebase/database": "^13.0.1"

my log>>>

![firerefdoterror](https://user-images.githubusercontent.com/57331729/142775437-ad657d1c-40c5-4dd8-bd31-283378d11ace.PNG)

will really appreciate it if i can get a swift resolution, thanks

**Comment by mikehardy** [comment]
@deflexable please open a new issue following issue template, ideally with an App.js that would reproduce if dropped in a https://stackoverflow.com/help/minimal-reproducible-example like that generated from https://github.com/mikehardy/rnfbdemo/blob/master/make-demo.sh - and since this appears to be an issue with a success report on the PR that closed it, a pointer to the docs indicating what you are trying to do is valid. Thanks

**Comment by saad9624-tribalScale** [comment]
firebase.database() Paths must be non-empty strings and can't contain ".", "#", "$", "[", or "]"

I am getting same error while sending "."

"@react-native-firebase/app": "^14.7.0",
"@react-native-firebase/database": "^14.7.0",

**Comment by mikehardy** [comment]
New issue please?

**Comment by chamaldesilva** [comment]
> New issue please?

The same issue is in react native.

firebase.database() Paths must be non-empty strings and can't contain ".", "#", "$", "[", or "]"

"@react-native-firebase/app": "^14.9.0",
"@react-native-firebase/database": "^14.9.0",

**Comment by NguyenDucThuanBKDN** [comment]
Hi,
We’re experiencing an issue related to an Invalid Path. Could you please check it for us? The log is as follows:
`default	18:26:28.168431+0700	Uncaught exception: InvalidPathValidation: (child:) Must be a non-empty string and not contain '.' '#' '$' '[' or ']'
(
	0   CoreFoundation                      0x000000019cfb2684 5A6C1F41-BF70-32F6-A1D6-5B894DD21362 + 968324
	1   libobjc.A.dylib                     0x000000019522fc80 objc_exception_throw + 60
	2   UnityFramework                      0x0000000111e157f0 firebaseJobsTroll + 6864
	3   UnityFramework                      0x0000000111dde0a8 f_b64_pton + 80416
	4   UnityFramework                      0x0000000113785910 _ZN8firebase8database8internal25DatabaseReferenceInternal5ChildEPKc + 160
	5   UnityFramework                      0x000000011376d268 _ZNK8firebase8database17DatabaseReference5ChildEPKc + 92
	6   UnityFramework                      0x0000000113799f54 Firebase_Database_CSharp_InternalDatabaseReference_Child__SWIG_0 + 144
	7   UnityFramework                      0x000000010ea1ec50 UnityFramework + 15526992
	8   UnityFramework                      0x000000010ea1eb60 UnityFramework + 155`

Many thanks!

**Comment by mikehardy** [comment]
@NguyenDucThuanBKDN with apologies, no one will check that for you, that comment looks like nothing like a standard issue report containing enough information to help. I suspect you are attempting to send a path through that is invalid though, just as the error says. Additionally 'UnityFramework' looks suspiciously like it is not a react-native-firebase issue.

**Comment by NguyenDucThuanBKDN** [comment]
Hi @mikehardy , 
Thank for your support.

Sorry, at the moment we’re not sure why the wrong path was passed in. From the error message, it seems to be an invalid path. However, I have one question:
if the path passed in is incorrect, why doesn’t it throw an exception or an error that we can handle, instead of terminating the app (crashing the app)?

**Comment by mikehardy** [comment]
> that comment looks like nothing like a standard issue report containing enough information to help. 

as well as...

>  Additionally 'UnityFramework' looks suspiciously like it is not a react-native-firebase issue.
